[migrate] Improved assertion

kunjsong01 opened the pull request:
made migrate name assertion more understandable as a result of Issue https://github.com/esbmc/esbmc/issues/720

kunjsong01 committed:
[migrate] Improved error reporting

  - replaced assert with msg.error and abort
  - added comment

mikhailramalho reviewed (changes requested):
I think when we want to print an error message, it should be:
1. print to cerr (using the message interface)
2. abort.

kunjsong01 commented:
ACK
will replace it by a proper error message printing.

fbrausse commented:
Do we want to print a message here for this clearly internal (opposed to usage-related) error? Why is the assertion and maybe an accompanying comment not enough?

mikhailramalho commented:
I think error messages are always useful, even if it's only for the user to
report the error to us.

On Wed, 1 Jun 2022, 11:47 Franz Brauße, ***@***.***> wrote:

> Do we want to print a message here for this clearly internal (opposed to
> usage-related) error? Why is the assertion and maybe an accompanying
> comment not enough?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/721#issuecomment-1143441224>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJHZDLZUWUOSW5CQO7MTVM45UJANCNFSM5XPDFBOA>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

fbrausse commented:
>> clearly internal

> always

Are you suggesting to favour unconditionally enabled checks and messages over `assert()` and `NDEBUG` generally?

mikhailramalho commented:
Personally, yes.

In ESBMC, I think we use asserts in a number of places and always release using RelWithDebInfo (-O2 -g).

lucasccordeiro commented:
What is the conclusion here?

fbrausse commented on src/util/migrate.cpp:
```diff
@@ -510,9 +510,9 @@ expr2tc sym_name_to_symbol(irep_idt init, type2tc type)
     thread_num = strtol(exmstr.c_str(), &endexmptr, 10);
-    assert(endexmptr != exmstr.c_str());
+    if(endatptr == atstr.c_str() || endexmptr == exmstr.c_str())
+      assert(!"Failed to mirgate symbol due to invalid identifier.");
```
Typo: migrate. If you're printing this, please also include the identifier.

fbrausse reviewed (commented):
For me a comment at the assertion would have sufficed, but appearently not-compiled-out checks and error messages to the user seem to find enough support. Fine with me here as this code is run relatively infrequently and the checks are not expensive. For things like #204 or similar however, I would suggest to keep assertions for performance reasons.

mikhailramalho committed:
Revert "[regression] Added new test cases and improved documentation"

This reverts commit df4e50bb49cf3dfcba6e1f9d9ffa11280f41eabb.

mikhailramalho committed:
Revert "[clang-cpp-frontend] Test case bring-up - Part III"

This reverts commit 5b78c70d30fcb7338d3a57825c4e5a90a96bf460.

mikhailramalho committed:
Revert "[regression] updated constructor test suite"

This reverts commit 26be7bf753b60ebf734acafff619857e77b8eaad.

mikhailramalho committed:
Revert "[clang-cpp-frontend] code clean-up"

This reverts commit c71c91d576bd33ae020dce48fb6d3cb16b5f7ed4.

mikhailramalho committed:
Revert "[clang-cpp-frontend] use push_back instead of indexing operator to add this pointer"

This reverts commit 7d9dbefb6656bb086b0d3ef287a93ba43d649ccb.

mikhailramalho committed:
Revert "[clang-cpp-frontend] Test case bring-up - Part V"

This reverts commit 7386646dfaee678037af202f274ee207b1c52e77.

mikhailramalho committed:
Revert "[clang-cpp-frontend] Test case bring-up - Part IV"

This reverts commit 5828ee12f6386d461783316813ca0a17a55e7eeb.

mikhailramalho committed:
Revert "[clang-c-frontend] split get_function_params"

This reverts commit ef4ae911e1866786acbd3bd6adaa4f0f3cc68a39.

mikhailramalho committed:
Revert "[regression] Disabled template test cases"

This reverts commit 17c20d460a44192f5a22e90d315cb32946ba7112.

mikhailramalho committed:
Revert "[clang-cpp-frontend] Test case bring-up - Part II"

This reverts commit 2d60dcb60dd0a1d6a2b2b156160f20e6c0c52204.

mikhailramalho committed:
Revert "[clang-cpp-frontend] Construction test cases"

This reverts commit eb96ff6298465c1cc805b1cb337c8582d1fa75e4.

mikhailramalho committed:
Revert "[clang-cpp-frontend] Test case bring-up - Part VI"

This reverts commit df0e687ebcf9006ea499e260c6b66be2e5b78724.

mikhailramalho committed:
Revert "[clang-cpp-frontend] Test case bring-up - Part I"

This reverts commit 12883e0ec90a6f347fd079b15fde7a6f3a2608a7.

kunjsong01 commented:
Based on the discussion above, I've done the following actions: 

1. Added a comment to make the assertion more understandable while avoiding additional checks (which might cause performance degradation)
2. Replaced the `assert` error reporting with our standard `msg` interface and abort.  

---
P.S. It's not the first time seeing error reporting using `assert`, e.g. https://github.com/esbmc/esbmc/pull/670/files. This is what makes me think that both `assert` and standard interfaces were acceptable for error reporting. But apparently they just exist the legacy code and should no longer be used.

lucasccordeiro commented:
@mikhailramalho and @fbrausse: Can I ask you whether you still have comments for this PR?

fbrausse commented on src/util/migrate.cpp:
```diff
@@ -510,9 +511,10 @@ expr2tc sym_name_to_symbol(irep_idt init, type2tc type)
     assert(endatptr != atstr.c_str());
     thread_num = strtol(exmstr.c_str(), &endexmptr, 10);
-    assert(endexmptr != exmstr.c_str());
+    assert(endatptr != atstr.c_str());
```
Please restore the original check here. No use checking the same thing twice ;)

kunjsong01 replied:
Good spot! 
sorry, it's a copy-paste error...

mikhailramalho reviewed (approved):
LGTM.
